### PR TITLE
Feat/merkle distributor

### PR DESCRIPTION
**Pull request by blackalpacaa**
## Description

- Add Merkle Distributor contract.

## Why?

- To allow users to claim reward on feature pool.

## ChangeLogs:
- Add MerkleDistributor contract

### CODE REVIEW

**Review by almost-llamas — APPROVED** [review]

File: contracts/6/token/MerkleDistributor.sol [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    claimedBitMap[claimedWordIndex] = claimedBitMap[claimedWordIndex] | (1 << claimedBitIndex);
+  }
+
+  function claim(
+    uint index,
+    address account,
+    uint amount,
+    bytes32[] calldata merkleProof
+  ) external override {
+    require(!isClaimed(index), 'MerkleDistributor: Drop already claimed.');
```
Comment (spicysquid168) [inline contracts/6/token/MerkleDistributor.sol:48]: Use `Contract::method:: reason` format for revert message

File: deploy/017_deploy_merkle_distributor.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    ░╚██╗████╗██╔╝███████║██████╔╝██╔██╗██║██║██╔██╗██║██║░░██╗░
+    ░░████╔═████║░██╔══██║██╔══██╗██║╚████║██║██║╚████║██║░░╚██╗
+    ░░╚██╔╝░╚██╔╝░██║░░██║██║░░██║██║░╚███║██║██║░╚███║╚██████╔╝
+    ░░░╚═╝░░░╚═╝░░╚═╝░░╚═╝╚═╝░░╚═╝╚═╝░░╚══╝╚═╝╚═╝░░╚══╝░╚═════╝░
+    Check all variables below before execute the deployment script
+    */
+    const MERKLE_ROOT = '0x54be7bbc9d4950685ec4b2fd98d18e092bbd690fff6aa18c3e301ec8653d88f1'; //merkle root ITAM week 1 testnet.
+    const FEATURE_TOKEN_ADDRESS = '0xd817BfBE43229134e7127778a96C0180e47c10B4'; // itam token address testnet
+
+    console.log(">> Deploying a Merkle distributor contract");
```
Comment (spicysquid168) [inline deploy/017_deploy_merkle_distributor.ts:19]: put a lot of enter before this console.log so when we deploy we clearly see that this is a parameters that needed to be updated

File: test/MerkleDistributor.test.ts [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+      const { claims: innerClaims, merkleRoot, tokenTotal } = parseBalanceMap({
+        [await alice.getAddress()]: 200,
+        [await bob.getAddress()]: 300,
+        [await catherine.getAddress()]: 250,
+      })
+      const MerkleDistributorContract = (await ethers.getContractFactory(
+        "MerkleDistributor",
+        deployer
+        )) as MerkleDistributor__factory;
```
Comment (spicysquid168) [inline test/MerkleDistributor.test.ts:57]: no tab please

File: test/MerkleDistributor.test.ts [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+        merkleDistributor = await MerkleDistributorContract.deploy(mockERC20.address, merkleRoot)
+        await merkleDistributor.deployed();
+        
+        merkleAsDeployer = MerkleDistributor__factory.connect(merkleDistributor.address, deployer)
+
+      expect(tokenTotal).to.eq('0x02ee') // 750
```
Comment (spicysquid168) [inline test/MerkleDistributor.test.ts:63]: change from fixed '0x02ee' to convert 750 to hex and pass this to expect

File: test/MerkleDistributor.test.ts [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+      claims = innerClaims
+      await mockTokenAsDeployer.mint(merkleDistributor.address, tokenTotal)
```
Comment (spicysquid168) [inline test/MerkleDistributor.test.ts:65]: change from mint to merkleDistriubtor address to mint to some EOA account and use that account to deposit to merkle distributor contract

File: test/MerkleDistributor.test.ts [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+    })
+    context('parseBalanceMap', () => {
```
Comment (spicysquid168) [inline test/MerkleDistributor.test.ts:67]: 'validate parseBalanceMap function'

File: test/MerkleDistributor.test.ts [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+      it('check the proofs is as expected', async () => {
```
Comment (spicysquid168) [inline test/MerkleDistributor.test.ts:68]: 'should generate correct proof'

File: test/MerkleDistributor.test.ts [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+          '0x947a7b7d06336aaaad02bfbe9ae36b7ad7b5d36a98931901e958544620016414',
+          '0xa1281640dd3f2f3e400a42e90527e508f5a7ee4286dff710c570775145ee0165'
+        ])
+        // Catherine
+        expect(claims[await catherine.getAddress()].amount).to.equal('0xfa')
+        expect(claims[await catherine.getAddress()].proof).to.deep.equal([
+          '0xd1df20cc2fcab841bf3870b019038437dbd4c8db2bff627a8b385ebcc951f3a6',
+        ])
+      })
+      it('all claim work exactly once', async () => {
```
Comment (spicysquid168) [inline test/MerkleDistributor.test.ts:87]: it should allow to claim only once

File: test/MerkleDistributor.test.ts [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+          )
+        }
+        expect(await mockTokenAsDeployer.balanceOf(merkleAsDeployer.address)).to.eq(0)
+      })
+      it(`all reward's distributed to correct person`, async () => {
+        for (let account in claims) {
+          const claim = claims[account]
+          await merkleAsDeployer.claim(claim.index, account, claim.amount, claim.proof)
+        }
+        expect(await mockTokenAsDeployer.balanceOf(await alice.getAddress())).to.eq('0xc8')
```
Comment (spicysquid168) [inline test/MerkleDistributor.test.ts:105]: change from hex to decimal

**Review by spicysquid168 — APPROVED** [review]

**Review by almost-llamas — APPROVED** [review]

**Pull request merged**